Refactor curriculum schema to version 2.0.0

fazzy12 opened the pull request:
- Updated curriculum schema to version 2.0.0, modifying the structure to establish a 1:1 relationship with persons and an array of resources.
- Removed the "personId" property from the curriculum schema.
- Added "resources" array within the curriculum schema to hold resource objects with "resourceId" and "topicId" properties.
- Updated the configuration file to version 2.0.0 and removed the index on "personId" accordingly.
- Updated the test data in curriculum.json to reflect the new schema structure, ensuring each curriculum object contains an array of resources.

fazzy12 committed:
Refactor curriculum schema to version 2.0.0

FlatBallFlyer commented on src/mongosh/config/curriculum.json:
```diff
@@ -1,11 +1,15 @@
+    {
+      "keys": {"resources": 1},
+      "options": {"unique": false}
+    },
```
why do we need this index?

FlatBallFlyer commented on src/mongosh/schemas/curriculum-2.0.0.json:
```diff
@@ -1,33 +1,49 @@
-		"_id"
+		"_id",
+		"personId",
+		"resources"
```
I think ``personID`` is no longer needed as ``_id`` will be the ``perseonId``
Be careful about requiring resources - when we first create a curriculum it will not have anything other than ``_id``.

fazzy12 replied:
okay got it.

FlatBallFlyer commented on src/mongosh/schemas/curriculum-2.0.0.json:
```diff
@@ -1,33 +1,49 @@
+					}
+				},
+				"required": ["resourceId", "topicId"]
+			}
```
A few things here:
- Be cautious about required properties, keep in mind auto-save features
- I think the resource object in the array should have a ``type`` attribute that is ``topic`` or ``adhoc`` 
- for "topic" type resources we have a resourceId, topicId
- for "adhoc" type resources we have a name, link
- all resource objects should have a review and roadmap property
- roadmap values should be enumerated to 
   - completed
   - now
   - next
   - later

this is a more complex structure and will require some additional knowledge. Research json schema and the ``oneOf`` construct.

FlatBallFlyer commented on src/mongosh/data/curriculum.json:
```diff
@@ -1,6 +1,10 @@
     }
-]
\ No newline at end of file
+]
```
Let's build more test data - examine the Topics test data for Topic/Resource ID's to use, and People data for PersonID values to re-use as curriculumId's.

FlatBallFlyer commented on src/mongosh/config/curriculum.json:
```diff
@@ -1,11 +1,15 @@
   "migrations": []
-}
\ No newline at end of file
+}
```
change personId to _id and let's rename roadmapGroup to just roadmap

FlatBallFlyer reviewed (changes requested):
Good start, I left some feedback about changes.

fazzy12 committed:
Refactor curriculum schema to version 2.0.0

fazzy12 commented:
- Removed 'personId' field from the curriculum-2.0.0.json schema as '_id' will serve as the personId.
- Adjusted required properties, considering auto-save features.
- Introduced 'type' attribute for resource objects, distinguishing between 'topic' and 'adhoc'.
- Defined attributes based on resource type: 'resourceId' and 'topicId' for 'topic' type, and 'name' and 'link' for 'adhoc' type.
- Ensured all resource objects include 'review' and 'roadmap' properties.
- Enumerated roadmap values as 'completed', 'now', 'next', and 'later'.
- Researched and implemented the 'oneOf' construct for complex data structures in the JSON schema.
- Generated additional test data aligning with Topics and People datasets.
